### PR TITLE
JsonDB storage service

### PR DESCRIPTION
**Pull request by cdjackson**
This is an implementation of a json storage service for initial comment. Its main advantage is that it allows a user to edit the files with a text editor or json editor and is easily human readable. It is also considerably faster than mapdb. It has the potential to be less robust, but given it can be edited, and backups easily maintained, this is considered to be a low risk (currently!).

This implements the StorageService as a Map, but persists the data to disk as json files. Each storage that is opened is persisted to a separate file. Data is only read from the disk on startup.

The system employs a delayed write mechanism (currently 500ms) to improve performance where there are multiple writes in a short time.  When the service is closed, all open services are written to disk so there should be no risk of loosing data unless the JVM crashes and the delay could be increased a little (probably still best to keep it low to reduce the chance of lost data due to a crashed JVM).

The service keeps backups of each table - currently up to 5 backups are maintained. Each time the json file is written, the old file is moved into a backup folder, and old files are removed from the backup folder.

Benchmark tests indicate that this is significantly faster than the mapdb storage service (20 to 30 times faster). With the pretty print option enabled the storage size is similar to the mapdb, and about 30% smaller if this is disabled. Currently I've left this enabled since the major advantage of this service is that the data can be edited easily so improving its readability at the expense of some disk space is considered acceptable.

**Benchmarks**

| Test | MapDB | JsonDB |
| --- | --- | --- |
| 2000 writes | 8.2 seconds | 0.2 seconds |
| 20000 writes | 80 seconds | 0.4 seconds |

### CODE REVIEW

**Comment by maggu2810** [comment]
The benchmark sounds interesting.

@pwolfart did IMHO a great job testing database corruption of the mapdb. If you are interested, have a look at
* 1455
* https://www.eclipse.org/forums/index.php?t=msg&th=1075047&goto=1731013&#msg_1731013

If your implementation is called to write very 499 ms it seems that the data is never written.
A very theoretical use case, but perhaps it should not be deferred all the time?

It is a nice feature that you can manually edit the storage on corruption, but I assume this is a feature for us / developers only. This is nothing that a normal user wants to do.
From my point of view the most important point is that we never need to fix a corrupt database because it does not corrupt at all or uses some backup strategy itself.

But IMHO we can merge such a storage backend (have not had a look at the code) and leave it to the product which one is chosen to use.

**Comment by hakan42** [comment]
Honestly, I would love to have such an (optional) storage system, especially during development but also while setting up a fresh system. I just love to see what makes a system work :smile:

**Comment by hfanieng** [comment]
For me as very experienced power this feature is welcome. Being beta-tester from the first version of OH2 it was a nightmare to reconfigure over 100 of items after the database-corruption.

**Comment by cdjackson** [comment]
> If your implementation is called to write very 499 ms it seems that the data is never written.

Interesting corner case - I'll take a look. I would guess in reality this nearly never happens, but it would be good to cover.
However - the data will still be ok - it will be written when the service is deactivated, so assuming there isn't a JVM crash, it will be fine in this case at the moment.

> but I assume this is a feature for us / developers only. This is nothing that a normal user wants to do.

No - I disagree. There have been a lot of discussions on the OH2 forum where people want to be able to edit the database while still having the ability to use the UI to configure the system.

**Comment by cdjackson** [comment]
I added a maximum deferred timer of 60 seconds, so the data will be written every 60 seconds if there was a pathological instance of a continuous write.

File: bundles/storage/org.eclipse.smarthome.storage.json/src/main/java/org/eclipse/smarthome/storage/json/JsonStorage.java [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+                    System.currentTimeMillis() + "-" + file.getName());
+            file.renameTo(rename);
+
+            FileOutputStream outputStream;
+            try {
+                outputStream = new FileOutputStream(file, false);
+                outputStream.write(s.getBytes());
+                outputStream.close();
+            } catch (Exception e) {
+                e.printStackTrace();
```
Comment (watou) [inline bundles/storage/org.eclipse.smarthome.storage.json/src/main/java/org/eclipse/smarthome/storage/json/JsonStorage.java]: logger?

File: bundles/storage/org.eclipse.smarthome.storage.json/src/main/java/org/eclipse/smarthome/storage/json/JsonStorage.java [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+        if (commitTimerTask != null) {
+            commitTimerTask.cancel();
+            commitTimerTask = null;
+        }
+
+        // Create the timer task
+        commitTimerTask = new CommitTimerTask();
+
+        // Start the timer
+        commitTimer.schedule(commitTimerTask, WRITE_DELAY);
```
Comment (watou) [inline bundles/storage/org.eclipse.smarthome.storage.json/src/main/java/org/eclipse/smarthome/storage/json/JsonStorage.java]: Could WRITE_DELAY and MAX_DEFERRED_PERIOD be made user configurable so users can decrease write cycles to flash memory?

Comment (cdjackson) [inline bundles/storage/org.eclipse.smarthome.storage.json/src/main/java/org/eclipse/smarthome/storage/json/JsonStorage.java]: Possibly - I did think about adding config services, but I wanted to create a PR first to see if this was something that was of interest in ESH or not before doing too much work on it.

**Comment by smar000** [comment]
> No - I disagree. There have been a lot of discussions on the OH2 forum where people want to be able to edit the database while still having the ability to use the UI to configure the system.

I fully agree with Chris on this. I think it would be invaluable to a lot of power users and not just developers.

**Comment by maggu2810** [comment]
Okay, perhaps it has been misspelled using "developers" and "users".
I assume the most of the ones that added comments here are using openHAB. So an open source product that has been set up by yourself.
The Eclipse SmartHome framework is also used by commercial products.
The users I talked about are the ones that used a commercial system.

I don't think such an user is willing to edit files on his system.
The framework should ensure, that data does not get corrupt.
We cannot provide a framework that intention is **also** commercial usage and rely on "fixable by editing storage files".

As stated above, I think it is a nice feature and could be merged to the tree (IMHO).
Then a product can choose to use it, or use another one.

**Comment by cdjackson** [comment]
If the need is to have something that has higher integrity, then it would be relatively simple to add extra checks to improve the integrity by reading back from disk, and automatically reverting to the previous backup.

**Comment by shorty707** [comment]
When will this be added to OpenHAB2 :-)
many people will be happy :)

File: bundles/storage/org.eclipse.smarthome.storage.json/about_files/LICENSE-2.0.txt [hunk 3]
```diff
@@ -0,0 +1,202 @@
+
```
Comment (maggu2810) [inline bundles/storage/org.eclipse.smarthome.storage.json/about_files/LICENSE-2.0.txt]: Which component of this PR is licensed using the APL2?

Comment (cdjackson) [inline bundles/storage/org.eclipse.smarthome.storage.json/about_files/LICENSE-2.0.txt]: Not sure how this ended up in the repo - I’ll remove it before this is merged.

File: bundles/storage/org.eclipse.smarthome.storage.json/src/main/java/org/eclipse/smarthome/storage/json/JsonStorage.java [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+        } catch (JsonSyntaxException | JsonIOException | FileNotFoundException e) {
+            logger.error("Error reading JsonDB from {}. Cause {}.", inputFile.getPath(), e.getMessage());
+            return null;
+        }
+    }
+
+    private File getBackupFile(int age) {
+        // Delete old backups
+        List<Long> fileTimes = new ArrayList<Long>();
+        File folder = new File(file.getParent() + File.separator + "backup");
```
Comment (kaikreuzer) [inline bundles/storage/org.eclipse.smarthome.storage.json/src/main/java/org/eclipse/smarthome/storage/json/JsonStorage.java]: Better define "backup" as a constant.

File: bundles/storage/org.eclipse.smarthome.storage.json/src/main/java/org/eclipse/smarthome/storage/json/JsonStorage.java [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+            // Rename the file for backup
+            File rename = new File(file.getParent() + File.separator + "backup",
+                    System.currentTimeMillis() + SEPARATOR + file.getName());
+            file.renameTo(rename);
+
+            FileOutputStream outputStream;
+            try {
+                outputStream = new FileOutputStream(file, false);
+                outputStream.write(s.getBytes());
+                outputStream.close();
```
Comment (kaikreuzer) [inline bundles/storage/org.eclipse.smarthome.storage.json/src/main/java/org/eclipse/smarthome/storage/json/JsonStorage.java]: either move this to finally or use try-with-resources

File: bundles/storage/org.eclipse.smarthome.storage.json/src/main/java/org/eclipse/smarthome/storage/json/JsonStorage.java [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+    }
+
+    public synchronized void deferredCommit() {
+        // Handle a maximum time for deferring the commit.
+        // This stops a pathological loop preventing saving
+        if (deferredSince != 0 && deferredSince < System.currentTimeMillis() - MAX_DEFERRED_PERIOD) {
+            commitDatabase();
+        }
+        if (deferredSince == 0) {
+            deferredSince = System.currentTimeMillis();
```
Comment (kaikreuzer) [inline bundles/storage/org.eclipse.smarthome.storage.json/src/main/java/org/eclipse/smarthome/storage/json/JsonStorage.java]: Better use System.nanoTime()

File: bundles/storage/org.eclipse.smarthome.storage.json/src/main/java/org/eclipse/smarthome/storage/json/JsonStorage.java [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+                    deleter.delete();
+                }
+            }
+        }
+    }
+
+    public synchronized void deferredCommit() {
+        // Handle a maximum time for deferring the commit.
+        // This stops a pathological loop preventing saving
+        if (deferredSince != 0 && deferredSince < System.currentTimeMillis() - MAX_DEFERRED_PERIOD) {
```
Comment (kaikreuzer) [inline bundles/storage/org.eclipse.smarthome.storage.json/src/main/java/org/eclipse/smarthome/storage/json/JsonStorage.java]: Better use System.nanoTime()

File: bundles/storage/org.eclipse.smarthome.storage.mapdb.test/src/main/java/org/eclipse/smarthome/storage/mapdb/MapdbStorageServiceBenchmark.java [hunk 8]
```diff
@@ -0,0 +1,100 @@
+package org.eclipse.smarthome.storage.mapdb;
```
Comment (kaikreuzer) [inline bundles/storage/org.eclipse.smarthome.storage.mapdb.test/src/main/java/org/eclipse/smarthome/storage/mapdb/MapdbStorageServiceBenchmark.java]: please add license header

File: extensions/persistence/org.eclipse.smarthome.persistence.h2sql/.project [hunk 9]
```diff
@@ -0,0 +1,33 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (kaikreuzer) [inline extensions/persistence/org.eclipse.smarthome.persistence.h2sql/.project]: looks like an accidential commit?

**Comment by cdjackson** [comment]
@kaikreuzer Hopefully this is updated to reflect all your comments.

File: bundles/storage/org.eclipse.smarthome.storage.json/src/main/java/org/eclipse/smarthome/storage/json/JsonStorageService.java [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This implementation of {@link StorageService} provides a mechanism to store
+ * data in Json files.
+ *
+ * @author Chris Jackson - Initial Contribution
+ */
+public class JsonStorageService implements StorageService, ManagedService {
```
Comment (kaikreuzer) [inline bundles/storage/org.eclipse.smarthome.storage.json/src/main/java/org/eclipse/smarthome/storage/json/JsonStorageService.java]: please do not use ManageService - use DS component configuration through activate() instead.

File: bundles/storage/org.eclipse.smarthome.storage.json/src/main/java/org/eclipse/smarthome/storage/json/JsonStorageService.java [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    @Override
+    public void updated(Dictionary<String, ?> properties) throws ConfigurationException {
+        if (properties == null || properties.isEmpty()) {
+            return;
+        }
+
+        if (properties.get(CFG_MAX_BACKUP_FILES) != null) {
+            maxBackupFiles = Integer.parseInt((String) properties.get(CFG_MAX_BACKUP_FILES));
```
Comment (kaikreuzer) [inline bundles/storage/org.eclipse.smarthome.storage.json/src/main/java/org/eclipse/smarthome/storage/json/JsonStorageService.java]: Integer.parseInt can throw RuntimeExceptions, which you need to handle.

File: bundles/storage/org.eclipse.smarthome.storage.json/src/main/java/org/eclipse/smarthome/storage/json/JsonStorage.java [hunk 12]
```diff
@@ -0,0 +1,340 @@
...
+        return values;
+    }
+
+    /**
+     * Deserializes and instantiates an object of type {@code T} out of the
+     * given JSON String. A special classloader (other than the one of the
+     * Json bundle) is used in order to load the classes in the context of
+     * the calling bundle.
+     *
+     * @param json
```
Comment (kaikreuzer) [inline bundles/storage/org.eclipse.smarthome.storage.json/src/main/java/org/eclipse/smarthome/storage/json/JsonStorage.java]: either add some text or remove those lines

File: bundles/storage/org.eclipse.smarthome.storage.json/OSGI-INF/jsonstorage.xml [hunk 13]
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!-- Copyright (c) 2014-2015 openHAB UG (haftungsbeschraenkt) and others. All rights reserved. This program and the accompanying 
+	materials are made available under the terms of the Eclipse Public License v1.0 which accompanies this distribution, and 
+	is available at http://www.eclipse.org/legal/epl-v10.html -->
+<scr:component xmlns:scr="http://www.osgi.org/xmlns/scr/v1.1.0" activate="activate" deactivate="deactivate"
+	immediate="true" name="org.eclipse.smarthome.storage.json">
+	<implementation class="org.eclipse.smarthome.storage.json.JsonStorageService" />
+
+    <property name="service.pid" type="String" value="org.eclipse.smarthome.storage.json" />
+	<property name="service.config.description.uri" type="String" value="system:json_storage" />
```
Comment (kaikreuzer) [inline bundles/storage/org.eclipse.smarthome.storage.json/OSGI-INF/jsonstorage.xml]: I cannot find a config description for system:json_storage.
Anyhow, I don't think those settings need to be user adjustable through the UI - so I'd suggest to remove those service.config properties here.

**Comment by cdjackson** [comment]
@kaikreuzer regarding the config definition - I missed committing the file. However, I do think it's useful for people to be able to change these settings. For example, people using SSD/USB for storage may want to increase the delays to reduce the write cycles and avoid corruption. Sure, they could do it with config files, but doesn't it make more sense to have it configurable through the UI if config is available?

**Comment by kaikreuzer** [comment]
ok for me.

**Comment by cdjackson** [comment]
@kaikreuzer this should be updated now I hope

File: bundles/storage/org.eclipse.smarthome.storage.json/OSGI-INF/jsonstorage.xml [hunk 14]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!-- Copyright (c) 2014-2015 openHAB UG (haftungsbeschraenkt) and others. All rights reserved. This program and the accompanying 
+	materials are made available under the terms of the Eclipse Public License v1.0 which accompanies this distribution, and 
+	is available at http://www.eclipse.org/legal/epl-v10.html -->
+<scr:component xmlns:scr="http://www.osgi.org/xmlns/scr/v1.1.0" activate="activate" deactivate="deactivate"
+	immediate="true" name="org.eclipse.smarthome.storage.json">
+	<implementation class="org.eclipse.smarthome.storage.json.JsonStorageService" />
+
+    <property name="service.pid" type="String" value="org.eclipse.smarthome.storage.json" />
```
Comment (kaikreuzer) [inline bundles/storage/org.eclipse.smarthome.storage.json/OSGI-INF/jsonstorage.xml]: could you replace the spaces by tabs?

File: bundles/storage/org.eclipse.smarthome.storage.json/src/main/java/org/eclipse/smarthome/storage/json/JsonStorageService.java [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+    }
+
+    protected void activate(ComponentContext cContext, Map<String, Object> properties) {
+        modified(properties);
+    }
+
+    protected void deactivate(ComponentContext cContext, int reason) {
+    }
+
+    public synchronized void modified(Map<String, Object> properties) {
```
Comment (kaikreuzer) [inline bundles/storage/org.eclipse.smarthome.storage.json/src/main/java/org/eclipse/smarthome/storage/json/JsonStorageService.java]: You didn't specify the existence of the modified method in your component xml. And you shouldn't as your service is not able to deal with changes after it has been activated. Hence you should simply move this code in here to activate() and remove the modified() method.

File: bundles/storage/org.eclipse.smarthome.storage.json/src/main/java/org/eclipse/smarthome/storage/json/JsonStorageService.java [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+    private final String CFG_WRITE_DELAY = "write_delay";
+    private final String CFG_MAX_DEFER_DELAY = "max_defer_delay";
+
+    private int maxBackupFiles = 5;
+    private int writeDelay = 500;
+    private int maxDeferredPeriod = 60000;
+
+    private final Map<String, JsonStorage<Object>> storageList = new HashMap<String, JsonStorage<Object>>();
+
+    public void activate() {
```
Comment (kaikreuzer) [inline bundles/storage/org.eclipse.smarthome.storage.json/src/main/java/org/eclipse/smarthome/storage/json/JsonStorageService.java]: You have two activate methods now. You should move the code from here to the newly introduced one.

File: bundles/storage/org.eclipse.smarthome.storage.json/src/main/java/org/eclipse/smarthome/storage/json/JsonStorageService.java [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+        }
+        File backup = new File(dbFolderName, "backup");
+        if (!backup.exists()) {
+            backup.mkdirs();
+        }
+        logger.debug("Json Storage Service: Activated.");
+    }
+
+    public void deactivate() {
+        // Since we're using a delayed commit, we need to write out any data
```
Comment (kaikreuzer) [inline bundles/storage/org.eclipse.smarthome.storage.json/src/main/java/org/eclipse/smarthome/storage/json/JsonStorageService.java]: You have two deactivate methods now. You should move the code from here to the newly introduced one.

File: features/karaf/src/main/feature/feature.xml [hunk 18]
```diff
@@ -1,13 +1,4 @@
 <?xml version="1.0" encoding="UTF-8"?>
```
Comment (kaikreuzer) [inline features/karaf/src/main/feature/feature.xml]: This file seems to be completely messed up - could you please check?

File: bundles/storage/org.eclipse.smarthome.storage.json/src/main/java/org/eclipse/smarthome/storage/json/JsonStorageService.java (hunk 15 above)
Comment (cdjackson) [inline bundles/storage/org.eclipse.smarthome.storage.json/src/main/java/org/eclipse/smarthome/storage/json/JsonStorageService.java]: Good point - interestingly, modified does get called ;).

**Comment by cdjackson** [comment]
Updated again...

File: bundles/storage/org.eclipse.smarthome.storage.json/src/main/java/org/eclipse/smarthome/storage/json/JsonStorageService.java [hunk 19]
```diff
@@ -0,0 +1,98 @@
...
+            if (properties.get(CFG_MAX_BACKUP_FILES) != null) {
+                maxBackupFiles = Integer.parseInt((String) properties.get(CFG_MAX_BACKUP_FILES));
+            }
+            if (properties.get(CFG_WRITE_DELAY) != null) {
+                writeDelay = Integer.parseInt((String) properties.get(CFG_WRITE_DELAY));
+            }
+            if (properties.get(CFG_MAX_DEFER_DELAY) != null) {
+                maxDeferredPeriod = Integer.parseInt((String) properties.get(CFG_MAX_DEFER_DELAY));
+            }
+        } catch (NumberFormatException nfe) {
```
Comment (kaikreuzer) [inline bundles/storage/org.eclipse.smarthome.storage.json/src/main/java/org/eclipse/smarthome/storage/json/JsonStorageService.java]: Better do a catch for every single value - as it is now, you will ignore write_delay and defer_delay parameters, if backup_files cannot be parsed, even if they are set correctly.

File: bundles/storage/org.eclipse.smarthome.storage.json/src/main/java/org/eclipse/smarthome/storage/json/JsonStorageService.java [hunk 20]
```diff
@@ -0,0 +1,98 @@
...
+            logger.error("NumberFormatException: {}", nfe.getMessage());
```
Comment (kaikreuzer) [inline bundles/storage/org.eclipse.smarthome.storage.json/src/main/java/org/eclipse/smarthome/storage/json/JsonStorageService.java]: Better do a warning like: "Value xxx is invalid for parameter yyy - using zzz as a default instead"

**Comment by cdjackson** [comment]
Logging updated...

**Comment by kaikreuzer** [comment]
Thanks for your patience :-)
Ok, I will now create a CQ and ask for approval by the IP team. So please refrain from doing any further updates to this branch!

**Comment by kaikreuzer** [comment]
Created CQ 12251.

**Comment by maggu2810** [comment]
CQ has been approved.

**Comment by kaikreuzer** [comment]
@cdjackson Travis fails with a compilation error on the JSONStorage tests - could you please check? If it needs a fix, please add it as a new commit to your branch (do not squash!) - thanks!

**Comment by cdjackson** [comment]
I've updated this to remove the benchmark 'test' which was causing the issue. I don't think it's a good idea to have it - certainly not as a test anyway.

**Comment by kaikreuzer** [comment]
Thanks @cdjackson, but you forgot to sign-off your second commit!

**Comment by cdjackson** [comment]
Sorry - updated...

**Comment by kaikreuzer** [comment]
Perfect, thanks!

**Comment by maggu2810** [comment]
Isn't it still the wrong license header?

**Pull request merged**